nomad/csi_endpoint.go

```diff
@@ -792,6 +792,242 @@ func (v *CSIVolume) checkpointClaim(vol *structs.CSIVolume, claim *structs.CSIVo
 	return nil
 }
 
+func (v *CSIVolume) Create(args *structs.CSIVolumeCreateRequest, reply *structs.CSIVolumeCreateResponse) error {
+
+	if done, err := v.srv.forward("CSIVolume.Create", args, args, reply); done {
+		return err
+	}
+
+	metricsStart := time.Now()
+	defer metrics.MeasureSince([]string{"nomad", "volume", "create"}, metricsStart)
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityCSIWriteVolume)
+	aclObj, err := v.srv.WriteACLObj(&args.WriteRequest, false)
+	if err != nil {
+		return err
+	}
+
+	if !allowVolume(aclObj, args.RequestNamespace()) || !aclObj.AllowPluginRead() {
+		return structs.ErrPermissionDenied
+	}
+
+	if args.Volumes == nil || len(args.Volumes) == 0 {
+		return fmt.Errorf("missing volume definition")
+	}
+
+	regArgs := &structs.CSIVolumeRegisterRequest{
+		Volumes:      args.Volumes,
+		WriteRequest: args.WriteRequest,
+	}
+
+	// This is the only namespace we ACL checked, force all the volumes to use it.
+	// We also validate that the plugin exists for each plugin, and validate the
+	// capabilities when the plugin has a controller.
+	for _, vol := range args.Volumes {
+		vol.Namespace = args.RequestNamespace()
+		if err = vol.Validate(); err != nil {
+			return err
+		}
+		plugin, err := v.pluginValidateVolume(regArgs, vol)
+		if err != nil {
+			return err
+		}
+		if !plugin.ControllerRequired {
+			return fmt.Errorf("plugin has no controller")
+		}
+		if err := v.controllerValidateVolume(regArgs, vol, plugin); err != nil {
+			return err
+		}
+
+		// NOTE: creating the volume in the external storage provider can't be
+		// made atomic with the registration, and creating the volume provides
+		// values we want to write on the CSIVolume in raft anyways. For now
+		// we'll block the RPC on the external storage provider so that we can
+		// easily return meaningful errors to the user, but in the future we
+		// should consider creating registering first and creating a "volume
+		// eval" that can do the plugin RPCs async
+		err = v.createVolume(vol, plugin)
+		if err != nil {
+			return err
+		}
```

Comment:
> Would it make sense to have the createVolume be in a different loop following validation? To reduce the risk of orphaned volumes. Also, would it make sense to persist the successfully created volumes to Raft?

So in other words we could have one loop that does the validation and then a second loop that does the `createVolume` + `raftApply` below? I was originally thinking to avoid making extra raft writes but given the time it takes to do a `createVolume` that's probably not a useful concern. Will do.

Reply:
I think we can break it few ways - depending on how we want to handle partial failures. e.g. can do

1. Loop to validate
2. Loop to create volumes
3. Apply to raft with all the successfully created volumes

Reply:
Done in d4f4add